File: py/horus/pb/notification_service/service_handler.py
import abc as _abc
from ...rpc.base_handler import BaseRpcHandler, RpcMessage, WebSocket

from ..notification_service.service_pb2 import LogMessageEvent, LogMessageRequest, ProfilingInfoEvent, SensorInfoEvent
from ..rpc_pb2 import DefaultSubscribeRequest, DefaultSubscribeResponse, DefaultUnsubscribeRequest, DefaultUnsubscribeResponse

class NotificationServiceHandler(BaseRpcHandler, metaclass=_abc.ABCMeta):
    """
    Service which receives notifications from services, storing and forwarding them.
    """

    service_full_name = "horus.pb.NotificationService"
    service_id = 2

    def __init__(self, ws: WebSocket):
        super().__init__(ws, {
            1: (LogMessageRequest, None, self.log_message),
            3: (DefaultSubscribeRequest, DefaultSubscribeResponse, self.subscribe),
            4: (DefaultUnsubscribeRequest, DefaultUnsubscribeResponse, self.unsubscribe),
        })

    @_abc.abstractmethod
    async def log_message(self, request: LogMessageRequest) -> None:
        """
        Logs a message, storing it and broadcasting it to subscribers.
        """
        raise NotImplementedError

    @_abc.abstractmethod
    async def subscribe(self, request: DefaultSubscribeRequest) -> DefaultSubscribeResponse:
        """
        Subscribes to new log and profiling events.

        This function has a default implementation in C++, but not in Python.
        """
        raise NotImplementedError

    @_abc.abstractmethod
    async def unsubscribe(self, request: DefaultUnsubscribeRequest) -> DefaultUnsubscribeResponse:
        """
        Unsubscribe following a call to `Subscribe()`.

        This function has a default implementation in C++, but not in Python.
        """
        raise NotImplementedError

class NotificationListenerServiceHandler(BaseRpcHandler, metaclass=_abc.ABCMeta):
    """
    Service which listens for notifications forwarded by the `NotificationService`.
    """

    service_full_name = "horus.pb.NotificationListenerService"
    service_id = 10

    def __init__(self, ws: WebSocket):
        super().__init__(ws, {
            1: (LogMessageEvent, None, self.notify_log_message),
            2: (ProfilingInfoEvent, None, self.notify_profiling_info),
            3: (SensorInfoEvent, None, self.notify_sensor_info),
        })

    @_abc.abstractmethod
    async def notify_log_message(self, request: LogMessageEvent) -> None:
        """
        Received when a log message is received by the `NotificationService`.
        """
        raise NotImplementedError

    @_abc.abstractmethod
    async def notify_profiling_info(self, request: ProfilingInfoEvent) -> None:
        """
        Received when a profiling info is received by the `NotificationService`.
        """
        raise NotImplementedError

    @_abc.abstractmethod
    async def notify_sensor_info(self, request: SensorInfoEvent) -> None:
        """
        Received when a sensor info is received by the `NotificationService`.
        """
        raise NotImplementedError

File: cpp/horus/pb/notification_service/service_handler.cpp
#include "horus/pb/notification_service/service_handler.h"

#include <utility>

#include "horus/future/resolved.h"

namespace horus {
namespace sdk {
namespace pb {

AnyFuture<pb::RpcMessage> NotificationServiceHandler::Handle(const RpcContext& context,
                                                             pb::RpcMessage&& request) noexcept(false) {
  const pb::RpcMessage owned_request{std::move(request)};
  switch (owned_request.method_id()) {
    case 1: {
      return DoDispatch(&NotificationServiceHandler::LogMessage, context, owned_request);
    }
    case 3: {
      return DoDispatch(&NotificationServiceHandler::Subscribe, context, owned_request);
    }
    case 4: {
      return DoDispatch(&NotificationServiceHandler::Unsubscribe, context, owned_request);
    }
    default: {
      throw horus_internal::UnknownRpcMethodError{};
    }
  }
}

AnyFuture<pb::DefaultSubscribeResponse> NotificationServiceHandler::Subscribe(const RpcContext& context,
                                                                              pb::DefaultSubscribeRequest&& request) noexcept(false) {
  static_cast<void>(pb::DefaultSubscribeRequest{std::move(request)});
  return ResolveWith(NotificationListenerServiceSubscribers().Add<pb::DefaultSubscribeResponse>(context));
}

AnyFuture<pb::DefaultUnsubscribeResponse> NotificationServiceHandler::Unsubscribe(const RpcContext& context,
                                                                                  pb::DefaultUnsubscribeRequest&& request) noexcept(false) {
  static_cast<void>(pb::DefaultUnsubscribeRequest{std::move(request)});
  return ResolveWith(NotificationListenerServiceSubscribers().Remove<pb::DefaultUnsubscribeResponse>(context));
}

AnyFuture<pb::RpcMessage> NotificationListenerServiceHandler::Handle(const RpcContext& context,
                                                                     pb::RpcMessage&& request) noexcept(false) {
  const pb::RpcMessage owned_request{std::move(request)};
  switch (owned_request.method_id()) {
    case 1: {
      return DoDispatch(&NotificationListenerServiceHandler::NotifyLogMessage, context, owned_request);
    }
    case 2: {
      return DoDispatch(&NotificationListenerServiceHandler::NotifyProfilingInfo, context, owned_request);
    }
    case 3: {
      return DoDispatch(&NotificationListenerServiceHandler::NotifySensorInfo, context, owned_request);
    }
    default: {
      throw horus_internal::UnknownRpcMethodError{};
    }
  }
}

}  // namespace pb
}  // namespace sdk
}  // namespace horus
